File: QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;

/**
 *
 * @author venrra
 */
public class Jugador implements Comparable{
    private String nombre;
    private boolean encarcelado;
    private Casilla casillaActual;
    private Sorpresa cartaLibertad;
    private ArrayList<TituloPropiedad> propiedades;
    private int saldo;

    Jugador(String nombre) {
        this.nombre = nombre;
        setEncarcelado(false);
        this.casillaActual = null;
        this.cartaLibertad = null;
        this.propiedades = new ArrayList<>();
        this.saldo = 7500;
    }
    
    boolean cancelarHipoteca(TituloPropiedad titulo){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    boolean comprarTituloPropiedad(){
        int costeCompra = this.casillaActual.getCoste();
        boolean comprado=false;
        if (costeCompra < this.saldo){
            
            TituloPropiedad titulo = this.casillaActual.asignarPropietario(this);
            
            this.propiedades.add(titulo);
            
            this.modificarSaldo(-costeCompra);
            
            comprado=true;
        }
        
        return comprado;
    }
    
    int cuantasCasasHoteles(){
        int cont_CasasHotel=0;
        for (TituloPropiedad i: this.propiedades)
            cont_CasasHotel += i.getNumCasas() + i.getNumHoteles();
        return cont_CasasHotel;
    }
    
    boolean deboPagarAlquiler(){
        boolean deboPagar = false;
        TituloPropiedad titulo = casillaActual.getTitulo();
        boolean esDeMiPropiedad = this.esdeMiPropiedad(titulo);

        if(!esDeMiPropiedad){
            boolean tienePropietario = titulo.tengoPropiedad();
            if(tienePropietario){
                boolean encarcelado = titulo.propietarioEncarcelado();
                if(!encarcelado){
                    boolean estaHipotecada = titulo.getHipotecada();
                    deboPagar = !esDeMiPropiedad & tienePropietario & !encarcelado & !estaHipotecada;
                }
            }
        }
        
        return deboPagar;
    }
    
    Sorpresa devolverCartaLibertad(){
        Sorpresa aux  = this.cartaLibertad;
        
        if(cartaLibertad != null){
            cartaLibertad = null;
        }
        
        return aux;
    }
    
    boolean edificarCasa(TituloPropiedad titulo){
        boolean edificada = false;
        int numCasas = titulo.getNumCasas();
       
        if(numCasas < 4){
            
            int costeEdificarCasa = titulo.getPrecioEdificar();
            boolean tengoSaldo = this.tengoSaldo(costeEdificarCasa);
            
            if(tengoSaldo){
                
                titulo.edificarCasa();
                this.modificarSaldo(-costeEdificarCasa);            
                edificada = true;
            }
        }
        
        return edificada;
    }
    
    boolean edificarHotel(TituloPropiedad titulo){
        boolean edificada = false;
        int numHoteles = titulo.getNumHoteles();
       
        if(numHoteles < 4){
            
            int costeEdificarHotel = titulo.getPrecioEdificar();
            boolean tengoSaldo = this.tengoSaldo(costeEdificarHotel);
            
            if(tengoSaldo){
                
                titulo.edificarHotel();
                this.modificarSaldo(-costeEdificarHotel);            
                edificada = true;
            }
        }
        
        return edificada;
    }
    
    private void eliminarDeMisPropiedades(TituloPropiedad titulo){
        propiedades.remove(titulo);
        titulo.setPropietario(null);
    } 
    
    private boolean esdeMiPropiedad(TituloPropiedad titulo){
        return this.propiedades.contains(titulo);
    }
    
    boolean estoyEnCalleLibre(){
        throw new UnsupportedOperationException("Sinsi implementar");
    }

    String getNombre() {
        return nombre;
    }

    boolean getEncarcelado() {
        return encarcelado;
    }

    Casilla getCasillaActual() {
        return casillaActual;
    }

    Sorpresa getCartalibertad() {
        return cartaLibertad;
    }

    ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }    

    public int getSaldo() {
        return saldo;
    }
    
    void hipotecarPropiedad(TituloPropiedad titulo){
        int costeHipoteca = titulo.hipotecar();
        this.modificarSaldo(costeHipoteca);
    }
    
    void irACarcel(Casilla casilla){
        this.setCasillaActual(casilla);
        this.setEncarcelado(true);
    }
    
    int modificarSaldo(int cantidad){
        this.saldo += cantidad;
        
        return getSaldo();
    }
    
    int obtenerCapital(){
        int capital = saldo;
        
        for (TituloPropiedad i: propiedades){
            capital += i.getPrecioCompra() + (i.getNumCasas() + i.getNumHoteles()) * i.getPrecioEdificar();
            if ( i.getHipotecada() )
                capital -= i.getHipotecaBase();
        }
        
        return capital;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean estadoHipoteca){
        ArrayList<TituloPropiedad> titulosHipoteca = new ArrayList<>();
        
        if (estadoHipoteca == true){
            for ( TituloPropiedad i : propiedades)
                if (i.getHipotecada() == true)
                    titulosHipoteca.add(i);
        }else{
            for ( TituloPropiedad i : propiedades)
                if (i.getHipotecada() == false)
                    titulosHipoteca.add(i);
        }
     
            
        return titulosHipoteca;
    }
    
    void pagarAlquiler(){
        
        int costeAlquiler = this.casillaActual.pagarAlquiler();
        
        this.modificarSaldo(-costeAlquiler);
    }
    
    void pagarImpuesto(){
        this.saldo -= this.casillaActual.getCoste();
    }

    void pagarLibertad(int cantidad){
        boolean tengoSaldo = this.tengoSaldo(cantidad);
        
        if(tengoSaldo){
            this.setEncarcelado(false);
            this.modificarSaldo(cantidad);
        }
    }
    
    public void setCartalibertad(Sorpresa carta) {
        this.cartaLibertad = carta;
    }

    public void setCasillaActual(Casilla casilla) {
        this.casillaActual = casilla;
    }  
    
    public void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }
    
    boolean tengoCartaLibertad(){
        return this.cartaLibertad != null;
    }
            
    private boolean tengoSaldo(int cantidad){
        return ( saldo > cantidad );
    }
    
    void venderPropiedad(Casilla casilla){
        TituloPropiedad titulo = casilla.getTitulo();
        
        this.eliminarDeMisPropiedades(titulo);
        
        int precioVenta = titulo.calcularPrecioVenta();
        
        this.modificarSaldo(precioVenta);        
    }
    
    @Override
    public String toString() {
        String s = "Jugador{" + "nombre=" + nombre + ", encarcelado=" + encarcelado + ", casillaActual=" + casillaActual + ", cartaLibertad=" + cartaLibertad + ", saldo=" + saldo + ", capital=" + this.obtenerCapital();
        if (propiedades.size() != 0)
            s += ", propiedades=" + propiedades ;
        s += '}' + "\n";
        
        return s;
    }

    @Override
    public int compareTo(Object otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return otroCapital - obtenerCapital();
    }
}

File: QytetetRuby/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module ModeloQytetet
  class Jugador
    def initialize(nombre)
      @nombre = nombre
      @encarcelado = false
      @casillaActual = nil
      @cartaLibertad = nil
      @propiedades = Array.new
      @saldo = 7500
    end
    
    attr_accessor :cartaLibertad, :casillaActual, :encarcelado
    attr_reader :nombre, :propiedades, :saldo 
    public :saldo
   
    public
    def cancelarHiteca(titulo)
      raise NotImplementedError
    end
    
    def comprarTituloPropiedad()
      costeCompra = @casillaActual.coste
      comprado = false
      
      if costeCompra < @saldo
        titulo = @casillaActial.asgnarPropietario(self)
        @propiedades<<(titulo)
        modificarSaldo(-costeCompra)
        comprado = true
      end
      
     return comprado
    end
    
    def cuantasCasasHotelesTengo()
      contCasasHoteles = 0     
      @propiedades.each do |i|
        contCasasHotele += i.numCasas + i.numHoteles
      end    
      return contCasasHotele
    end
    
    def deboPagarAlquiler()
       deboPagar = false
       titulo = @casillaActual.titulo
       miPropiedad = esDeMiPropiedad(titulo)
       
      if (!miPropiedad)
        tienePropietario = titulo.tengoPropieda
        if (tienePropietario)
          encarcelado = titulo.propietarioEncarcelado
          if(!encarcelado)
            estaHipotecada = titulo.hipotecada
            deboPagar = !miPropiedad & tienePropietario & !encarcelado & !estaHipotecada
          end
        end
      end
      
      return deboPagar
    end
    
    def devolverCartaLibertad()
      aux = @cartaLibertad
      
      if (@cartaLibertad != nil)
        @cartaLibertad = nil
      end
      
      return aux
    end
    
    def edificarCasa(titulo)
      edificada = false
      numCasas = titulo.numCasas
      
      if numCasas < 4
        costeEdificar = titulo.precioEdificar
        tengo = tengoSaldo(costeEdificar)
        
        if tengo
          
          titulo.edificarCasa
          modificarSaldo(-costeEdificarCasa)
          edificada = true
          
        end
      end
      
      return edificada
    end
    
    def edificarHotel(titulo)
      edificada = false
      numHoteles = titulo.numHoteles
      
      if numCasas < 4
        costeEdificar = titulo.precioEdificar
        tengo = tengoSaldo(costeEdificar)
        
        if tengo
          
          titulo.edificarHotel
          modificarSaldo(-costeEdificarHotel)
          edificada = true
          
        end
      end
      
      return edificada      
    end
    
    private
    def eliminarDeMisPropiedades(titulo)
        @propiedades.remove(titulo)
         titulo.propietario(null)
    end
    
    def esDeMiPropiedad(titulo)
     
      for i in 0..@propiedades.size   
        if(@propiedades[i] == titulo)
          return true
        end
      end
      
      return false
    end
    
    public
    def estoyEnCalleLibre()
      raise NotImplementedError 
    end
    
    def hipotecarPropiedad(titulo)
      costeHipoteca = titulo.hipotecar
      modificarSaldo(costeHipoteca)
    end
    
    def irACarcel(casilla)
      @casillaActual = casilla
      @encarcelado = true
    end
    
    def modificarSaldo(cantidad)
      @saldo +=cantidad
      
      return @saldo
    end
    
    def obtenerCapital()
      capital = @saldo
      
      @propiedades.each do |i|
        capital += i.precioCompra + (i.numCasas + i.numHoteles) * i.precioEdificar;
        if(i.hipotecada)
          capital -= i.hopotecaBase
        end
      end
      
      return capital
    end
    
    def obtenerPropiedades(estadoHipoteca)
      titulosHipoteca = Array.new
      
      if(estadoHipoteca == true)
        @propiedades.each do |i|
          if i.hipotecada == true
            titulosHipoteca<<i
          end
        end
      else
        @propiedades.each do |i|
          if i.hipotecada == false
            titulosHipoteca<<i
          end
        end
      end
    end
         
    def pagarAlquiler()
      costeAlquiler = @casillaActual.pagarAlquiler
      modificarSaldo(-costeAlquiler)
    end
    
    def pagarImpuesto()
      @saldo -=  @casillaActual.coste
    end
    
    def pagarLibertad(cantidad)
      tengoSaldo = tengoSaldo(cantidad)
      
      if tengoSaldo 
        encarcelado(false)
        modificarSaldo(cantidad)
      end
    end
    
    def tengoCartaLibertad()
      return @cartaLibertad != nil
    end
    
    private
    def tengoSalgo(cantidad)
      return (@saldo>cantidad)
    end
    
    public
    def venderPropiedad(casilla)
      titulo = casilla.titulo
      eliminarDeMisPropiedades(titulo)
      precioVenta = titulo.calcularPrecioVenta
      modificarSaldo(precioVenta)
    end
    
    public
    def to_s
      "Jugador: \n nombre: #{@nombre}  \n encarcelado: #{@encarcelado} \n casilla actual: #{@casillaActual} \n carta libertad:#{@cartaLibertad} \n jugadores: \n propiedades: #{@propiedades.to_s} \n saldo: \n  #{@saldo} \n capital #{self.obtenerCapital}}"
    end
    
    def <=>(otroJugador)
      otroJugador.obtenerCapital <=> obtenerCapital
    end
  end
end
